contracts/d3vent.sol

```diff
@@ -0,0 +1,165 @@
+    // and so prevent the organiser withdrawal, and can claim back their funds
+    // a period up to which a joiner can unjoin and get a refund. after a certain point it's too late
+    // this should be expressed as a period before event start time. prevet organiser simply bringing the event forward and 
+    // being able rug pull / be sure prevent dirty tricks. e.g. make sure unjoining can't be shortened below a certain period
+    // need proof of personhood otherwise susceptible sybil attack from both sides
+
+    // add support for worldcoin plus another wallet
+
+    // access control to allow end user owner only. trustless or also admins. are there any organisational admins?
+
+    // should created events be immutable or setters for most things so they can be changed. what should be ammendable.
+    // if changes allowed, should have 1 setter function with all changeable settings, or setter for each
+    // organiser might reasonablely need ot change a event date or change a URI.
+    // change: organiser, multiple organisers, event name, uri, datetime, price capacity
+    // allow set different address as organiser at create time or able to change after
+    // create event: consider capacity of zero is unlimited or throw error?
+    // being able to change the event date does present possibilities of gaming the system and if allowed there should be restrictions
+    // if organiser change change dates, joiners should be allowed to unjoin and get their funds back.
+    // workflow questions: should newly created even have eventActive as param or always true/false
+    // should there be a way to cancel an event - how should that be represented
+    // should we have a cancelled / finished status
+
+    // create event organiser always msg.sender? setOrgnaiser function to change?
+    // should event have a closed/finished dateTime?
+    // should joiner address always be msg.sender or allow to join another address?
+    
+    // don't think there's any need to zero out joiner balances after organiser withdrawal, but could add a flag for added safety
+    // would be too gassy iterating through a large number of joiners
+    // 
+    
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity ^0.8.7;
+
+contract d3vent {
+    event CreatedEvent(uint indexed eventId, uint indexed eventDate, string indexed eventName);
+    event Withdrawal(uint indexed eventId, address indexed organiser, uint balance);
+    event JoinableSet(uint indexed eventId, bool isJoinable);
+       
+    uint eventIds;
+    uint immutable withdrawalBuffer;
+
+    //uint immutable unjoinBuffer; // not implemented
+    
+    // not implemented
+    // struct user {
+    //     address userAddress;
+    //     event_[] eventIds;  // do we need quick lookups and/or able to iterate eventIds?
+    // }
+
+    struct event_ {
+        address organiser;
+        uint id;
+        string name;
+        string uri;
+        uint dateTime;
+        uint price;
```

Comment:
A variable time should be added to define when will the event start(like 10 AM or 5 PM)

Reply:
it doesn't make sense to do that in the contract. that's a front end thing

Reply:
Suppose there are 100 events that will happen in the future. How can we store timings of that many event in the frontend if they're going to happen at different time

Reply:
I don't understand the question
The date and time of each event is stored in the event's dateTime field.
The frontend simply reads that, and has the job of translating dateTime <-> human readable date & time with each get and set.

Reply:
Okay so the workflow is going to look something like we will ask the organiser to provide a date and time for the event. Then we'll convert it into epoch format and then we'll be store it on chain. Right?

Reply:
My bad I thought that variable date time will just store plain date without converting it into epoch